### PR TITLE
Add embeddings_ layer and supporting utility functions

### PR DESCRIPTION
**Pull request by Cydral**
This PR introduces the `embeddings`_ layer class and adds two new utility functions, `embeddings()` and `embeddings_gradient()`, to the tensor_tools.h file. These additions enhance neural network capabilities, particularly for working with LLM-type networks.

### Key changes:

- Implemented embeddings_ layer: This new layer class allows for efficient token embedding in neural network architectures.
- Added embeddings function to tensor_tools.h: This utility function projects tokens from an input tensor into embeddings, facilitating efficient lookup and transformation of token representations.
- Added embeddings_gradient function to tensor_tools.h: This function computes and updates gradients for embeddings, supporting learning rate adjustment and optional frequency-based scaling. It's designed for thread-safe parallel processing.
- Included detailed function declarations with requires/ensures contracts in tensor_tools.h, providing clear documentation on usage and behavior of the utility functions.

### CODE REVIEW

**Comment by davisking** [comment]
Is this PR ready for review?  It's got conflicts with git master.  So not sure if I should be reviewing it yet or not.

**Comment by Cydral** [comment]
> Is this PR ready for review? It's got conflicts with git master. So not sure if I should be reviewing it yet or not.

Yes, you can go. I hope that these conflicts, which come (as I mentioned in another thread) from the fact that I've created several branches from my fork to work in parallel on the new Dlib layers, won't overwrite useful changes for certain layers. That said, and if there are compilation problems, I'll obviously fix the few problems, but all the layers currently released are ready for review and integration into the master branch. Thanks in advance.

File: dlib/cuda/tensor_tools.h [hunk 1]
```diff
@@ -2050,6 +2050,78 @@ namespace dlib { namespace tt
...
+                        r (0 <= r < dest.nr()),
+                        c (0 <= c < dest.nc()):
+                - Let token_idx = static_cast<unsigned long>(src(s,k,r,0))
+                - If token_idx < embs.num_samples():
+                    - #dest(s,k,r,c) = embs(token_idx, c, 0, 0)
+                - Else:
+                    - #dest(s,k,r,c) = 0
+            - The function iterates over all elements of src and populates dest accordingly.
+            - If a token index in src is out of range (>= embs.num_samples()),
+            the corresponding embedding in dest is filled with 0's.
```
Comment (davisking) [inline dlib/cuda/tensor_tools.h:2085]:
```suggestion
              the corresponding embedding in dest is filled with 0's.
```

**Comment by davisking** [comment]
Nice, this is cool.  Thanks for the PR :)
